File: week4/java/ScoreOfParentheses_856.java
class Solution {
    public int scoreOfParentheses(String S) {
        
        Stack<Integer> stack = new Stack<>();
        // initial score
        int currentScore = 0;
        
        for (char c : S.toCharArray()) {
            if (c == '(') {
                stack.push(currentScore);
                // new score
                currentScore = 0;
            } else {
                // update current score
                currentScore = Math.max(currentScore * 2, 1) + stack.pop();
            }
        }
        return currentScore;
    }
}

File: week4/python/ScoreOfParentheses_856.py
class Solution:
    def scoreOfParentheses(self, S: str) -> int:
        stack = []
        score = 0
        for char in S:
            if char == '(':
                stack.append(score)
                score = 0
            else:
                score = max(score * 2, 1) + stack.pop()
                
        return score
